### PR TITLE
Implementation of todo list

### PR DESCRIPTION
**Pull request by SanderBreivik**
Implements #2

Adds a Todo list where users can add, edit, delete and complete tasks. 
Completed tasks are at the bottom of the list, and a progress bar shows the progress of completion.

### CODE REVIEW

File: components/Todo/TodoScreen.js [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            autoFocus
+          />
+        </KeyboardAvoidingView>
+      </View>
+    );
+  }
+}
+
+class Task {
+
```
Comment (Artorp) [inline components/Todo/TodoScreen.js:179]: Should be called something like TaskPersistence

File: components/Todo/TodoScreen.js [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    } catch (e) {
+      console.error(e);
+    }
+  };
+
+  /**
+   * This method adds a new task, updates the state, and saves it to
+   * storage.
+   * @returns {Promise<void>}
+   */
```
Comment (Artorp) [inline components/Todo/TodoScreen.js:225]: missing method?

**Review by Artorp — CHANGES_REQUESTED** [review]
Looks great! See comments in code.

File: components/Todo/TodoScreen.js [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  /**
+   * Returns tasks (if any) and updates state
+   *
+   * @returns {Promise<void>} the return promise can be ignored
+   */
+  retrieveTasks = async () => {
+    try {
+      const taskKeysJson = await AsyncStorage.getItem("TASKKEYS") || "[]";
```
Comment (Artorp) [inline components/Todo/TodoScreen.js]: Maybe move persistence logic to the Task / TaskPersistence class?

**Review by Artorp — CHANGES_REQUESTED** [review]

File: components/Todo/TodoItem.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
+import {StyleSheet, Text, View, TextInput, Platform} from "react-native";
+import Icon from "react-native-vector-icons/Ionicons";
+import React, {Component} from "react";
+
+
+const isAndroid = Platform.OS === "android";
+export default class TodoItem extends Component {
+
+  handleToggleComplete = () => {
+    console.log(this.props);
```
Comment (Artorp) [inline components/Todo/TodoItem.js]: Remove debug log

File: components/Todo/TodoScreen.js [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+      const taskKeys = JSON.parse(taskKeysJson);
+      if (taskKeys.length > 0) {
+        const tasksResult = await AsyncStorage.multiGet(taskKeys);
+        console.log("taskRes:", tasksResult);
```
Comment (Artorp) [inline components/Todo/TodoScreen.js]: Remove debug log

File: components/Todo/TodoScreen.js [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+        const tasks = tasksResult
+          .map(keyvalue => keyvalue[1])
+          .map(objJson => JSON.parse(objJson));
+        this.setState({tasks: tasks});
+        console.log("tasks:", tasks);
```
Comment (Artorp) [inline components/Todo/TodoScreen.js]: Remove debug log

**Review by Artorp — CHANGES_REQUESTED** [review]
Should probably remove debug log statements

File: components/Todo/TodoScreen.js (hunk 3 above)
Comment (SanderBreivik) [reply components/Todo/TodoScreen.js]: How can i do that when i set the state?

Comment (Artorp) [reply components/Todo/TodoScreen.js]:
```js
const tasks = await Task.getAll();
setState({tasks});
```

Comment (SanderBreivik) [reply components/Todo/TodoScreen.js]: Then i'll have to make componentDidMount async, is that allowed?

Comment (Artorp) [reply components/Todo/TodoScreen.js]: No need, you're already calling retrieveTasks which is async

File: components/Todo/TodoScreen.js [hunk 7]
```diff
@@ -31,34 +31,17 @@ export default class TodoScreen extends Component {
     };
   }
 
-  componentDidMount() {
+  async componentDidMount() {
     // Fetch tasks from storage, if any.
-    this.retrieveTasks();
-  }
+    const tasks = await TaskPersistence.retrieveTasks();
+    this.setState({tasks: tasks});
```
Comment (Artorp) [inline components/Todo/TodoScreen.js]: Should set state only if tasks is defined, or make TaskPersistence.retrieveTasks() always return a list

File: components/Todo/TodoScreen.js [hunk 8]
```diff
@@ -234,13 +217,27 @@ class Task {
...
+    try {
+      const taskKeysJson = await AsyncStorage.getItem("TASKKEYS") || "[]";
+      const taskKeys = JSON.parse(taskKeysJson);
+      if (taskKeys.length > 0) {
+        const tasksResult = await AsyncStorage.multiGet(taskKeys);
+        const tasks = tasksResult
+          .map(keyvalue => keyvalue[1])
+          .map(objJson => JSON.parse(objJson));
+        return tasks;
+      }
```
Comment (Artorp) [inline components/Todo/TodoScreen.js:236]: See above comment in componentDidMount, either create a undefined check in componentDidMount, or make this method return an empty array

**Review by Artorp — CHANGES_REQUESTED** [review]

File: components/Todo/TodoScreen.js (hunk 8 above)
Comment (SanderBreivik) [reply components/Todo/TodoScreen.js:236]: Solved by check in componentDidMount, thanks for pointing out the bug! 👍

**Review by Artorp — APPROVED** [review]

**Pull request merged**